foyer/atomtyper.py

```diff
@@ -11,42 +13,54 @@ def find_atomtypes(structure, forcefield, max_iter=10):
 
     Parameters
     ----------
-    topology : simtk.openmm.app.Topology
+    topology : parmed.Structure or TopologyGraph
         The topology that we are trying to atomtype.
     forcefield : foyer.Forcefield
         The forcefield object.
     max_iter : int, optional, default=10
         The maximum number of iterations.
 
     """
-    typemap = {atom.idx: {'whitelist': set(), 'blacklist': set(), 
-        'atomtype': None} for atom in structure.atoms}
+    topology_graph = structure
+
+    if isinstance(structure, pmd.Structure):
+        topology_graph = TopologyGraph.from_parmed(structure)
+
+    typemap = {
+        atom_index: {
+            'whitelist': set(),
+            'blacklist': set(),
+            'atomtype': None
+        } for atom_index in topology_graph.atoms(data=False)
+    }
 
     rules = _load_rules(forcefield, typemap)
 
     # Only consider rules for elements found in topology
     subrules = dict()
 
     system_elements = set()
-    for a in structure.atoms:
+    for _, atom_data in topology_graph.atoms(data=True):
         # First add non-element types, which are strings, then elements
-        if a.name.startswith('_'):
-            if a.name in forcefield.non_element_types:
-                system_elements.add(a.name)
+        name = atom_data.name
+        atomic_number = atom_data.atomic_number
```

Comment:
This line is why I was asking about how non-element atom types handle `atomic_number`.